<a href="https://colab.research.google.com/github/moridin04/CCDATSCL_COM221ML_EXERCISES/blob/main/Exercise3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise 3

<img src="https://img.itch.zone/aW1nLzM0MzUxOTUuanBn/original/FuMxog.jpg" />

In this exercise, you will perform EDA on the popular Pokémon dataset, which contains detailed information about hundreds of Pokémon, including their stats, types, generation, and whether they are legendary.

The Pokémon dataset is ideal for practicing EDA because it includes a mix of numerical features (such as Attack, Defense, Speed, HP) and categorical features (such as Type 1, Type 2, and Generation). This combination allows you to apply a wide range of EDA techniques, including summary statistics, data visualizations, grouping and aggregation, correlation analysis, and comparisons across categories.

Throughout the exercise, you will explore questions such as:
- Which Pokémon tend to have the highest or lowest stats?
- How do different Pokémon types compare in terms of strength or defense?
- Do Legendary Pokémon differ significantly from non-Legendary ones?
- How are various stats distributed across the entire dataset?
- Are there relationships or trade-offs between certain attributes (e.g., Attack vs. Defense)?

In [1]:
import kagglehub
import os
import pandas as pd

In [2]:
# Download latest version
path = kagglehub.dataset_download("abcsds/pokemon")
print("Path to dataset files:", path)

100%|██████████| 14.9k/14.9k [00:00<00:00, 31.3MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/abcsds/pokemon/versions/2


In [3]:
if os.path.isdir(path):
  print(True)

contents = os.listdir(path)
contents

mydataset = path + "/" + contents[0]
mydataset


df = pd.read_csv(mydataset)

True



## 1: Data Understanding (4 pts)

1. Display the first 10 rows.

In [4]:
df.head(10)

#                       Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1                  Bulbasaur  Grass  Poison    318  45      49       49   
1  2                    Ivysaur  Grass  Poison    405  60      62       63   
2  3                   Venusaur  Grass  Poison    525  80      82       83   
3  3      VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4                 Charmander   Fire     NaN    309  39      52       43   
5  5                 Charmeleon   Fire     NaN    405  58      64       58   
6  6                  Charizard   Fire  Flying    534  78      84       78   
7  6  CharizardMega Charizard X   Fire  Dragon    634  78     130      111   
8  6  CharizardMega Charizard Y   Fire  Flying    634  78     104       78   
9  7                   Squirtle  Water     NaN    314  44      48       65   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False  
5       80       65     80           1      False  
6      109       85    100           1      False  
7      130       85    100           1      False  
8      159      115    100           1      False  
9       50       64     43           1      False

2. Show dataset shape.

In [6]:
df.shape

(800, 13)

3. Show all columns and its data types.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp. Atk     800 non-null    int64 
 9   Sp. Def     800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


4. Identify which columns contain missing values.

In [9]:
df.isnull().sum()

#               0
Name            0
Type 1          0
Type 2        386
Total           0
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

## 2. Summary Statistics (4 pts)

1. Generate `df.describe()`.

In [10]:
df.describe()

#      Total          HP      Attack     Defense     Sp. Atk  \
count  800.000000  800.00000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750  435.10250   69.258750   79.001250   73.842500   72.820000   
std    208.343798  119.96304   25.534669   32.457366   31.183501   32.722294   
min      1.000000  180.00000    1.000000    5.000000    5.000000   10.000000   
25%    184.750000  330.00000   50.000000   55.000000   50.000000   49.750000   
50%    364.500000  450.00000   65.000000   75.000000   70.000000   65.000000   
75%    539.250000  515.00000   80.000000  100.000000   90.000000   95.000000   
max    721.000000  780.00000  255.000000  190.000000  230.000000  194.000000   

          Sp. Def       Speed  Generation  
count  800.000000  800.000000   800.00000  
mean    71.902500   68.277500     3.32375  
std     27.828916   29.060474     1.66129  
min     20.000000    5.000000     1.00000  
25%     50.000000   45.000000     2.00000  
50%     70.000000   65.000000     3.00000  
75%     90.000000   90.000000     5.00000  
max    230.000000  180.000000     6.00000

2. Get mean, median, and mode of Attack.

In [11]:
df['Attack'].mean()

np.float64(79.00125)

In [12]:
df['Attack'].median()

75.0

In [13]:
df['Attack'].mode()

0    100
Name: Attack, dtype: int64

3. Compute 25th and 75th percentiles for HP.

In [14]:
print(df['HP'].quantile(0.25))
print(df['HP'].quantile(0.75))


50.0
80.0


4. Compute standard deviation and variance of Speed.

In [15]:
df['Speed'].std()

29.06047371716149

In [17]:
df['Speed'].var()

844.5111326658338

## 3. Filtering & Selection `(7 pts)`

Select all Pokémon with Attack > 100.

In [19]:
df.query('Attack > 100')

#                       Name   Type 1  Type 2  Total   HP  Attack  \
7      6  CharizardMega Charizard X     Fire  Dragon    634   78     130   
8      6  CharizardMega Charizard Y     Fire  Flying    634   78     104   
12     9    BlastoiseMega Blastoise    Water     NaN    630   79     103   
19    15      BeedrillMega Beedrill      Bug  Poison    495   65     150   
39    34                   Nidoking   Poison  Ground    505   81     102   
..   ...                        ...      ...     ...    ...  ...     ...   
793  717                    Yveltal     Dark  Flying    680  126     131   
796  719        DiancieMega Diancie     Rock   Fairy    700   50     160   
797  720        HoopaHoopa Confined  Psychic   Ghost    600   80     110   
798  720         HoopaHoopa Unbound  Psychic    Dark    680   80     160   
799  721                  Volcanion     Fire   Water    600   80     110   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
7        111      130       85    100           1      False  
8         78      159      115    100           1      False  
12       120      135      115     78           1      False  
19        40       15       80    145           1      False  
39        77       85       75     85           1      False  
..       ...      ...      ...    ...         ...        ...  
793       95      131       98     99           6       True  
796      110      160      110    110           6       True  
797       60      150      130     70           6       True  
798       60      170      130     80           6       True  
799      120      130       90     70           6       True  

[170 rows x 13 columns]

Select all Pokémon whose primary type (Type 1) is "Fire".

In [27]:
df.

ValueError: expr must be a string to be evaluated, <class 'bool'> given

Select all Pokémon that are Legendary.

In [ ]:
# put your answer here

Select all Pokémon that are Generation 1 AND Legendary.

In [ ]:
# put your answer here

Select all Pokémon that are Water type OR Grass type.

In [ ]:
# put your answer here

Select all Pokémon that are Fire type AND Attack > 120.

In [ ]:
# put your answer here

Select all Pokémon whose type is in this list:
`["Dragon", "Ghost", "Dark"]`.

In [ ]:
# put your answer here

## 4. Categorical Exploration `(9 pts)`

Find the number of Pokémon per primary type.

In [ ]:
# put your answer here

Find the number of Pokémon per generation.

In [ ]:
# put your answer here

Which type appears the most? Which appears the least?

In [ ]:
 # put your answer here

How many unique primary types (Type 1) exist?

In [ ]:
 # put your answer here

How many unique secondary types (Type 2) exist?

In [ ]:
 # put your answer here


Which primary types have the most dual-type combinations?

In [ ]:
 # put your answer here

Which type has the highest mean Attack?

In [ ]:
# put your answer here

Which type has the lowest mean Defense?

In [ ]:
# put your answer here


Which generation has the highest average Speed?

In [ ]:
# put your answer here

## 5. Groupby & Aggregation `(13 pts)`

Compute the average Attack per primary type.

In [ ]:
# put your answer here

Compute the maximum HP per generation.


In [ ]:
# put your answer here

Compute the total number of Pokémon per primary type.

In [ ]:
# put your answer here

For each type, compute:

- mean Attack

- mean Defense

- mean Speed

In [ ]:
# put your answer here

For each generation, compute:

- count

- mean Total

- number of Legendary Pokémon (hint: use sum on Boolean)

In [ ]:
# put your answer here

Which type combination (Type 1 + Type 2) has the highest average Attack?

In [ ]:
# put your answer here

Which generation has the highest proportion of Legendary Pokémon?

In [ ]:
# put your answer here


Which primary type has the largest variance in HP?

In [ ]:
# put your answer here

Which primary type has the highest median Speed?

In [ ]:
# put your answer here

Group Pokémon by whether they are Legendary or not. Compare:

- mean Total

- mean Attack

- mean Defense

- mean Speed

In [ ]:
# put your answer here

Show the top 5 strongest types by mean Total.

In [ ]:
# put your answer here

Rank generations by their average Attack.

In [ ]:
# put your answer here

Show the top 10 fastest Pokémon using nlargest(10, "Speed").

In [ ]:
# put your answer here